torch/csrc/jit/pickler.cpp

```diff
@@ -0,0 +1,391 @@
+#include <torch/csrc/jit/pickler.h>
+
+namespace torch {
+namespace jit {
+
+using ::c10::IValue;
+
+const std::vector<char>& Pickler::stack() {
+  return stack_;
+}
+
+void Pickler::start() {
+  push<OpCode>(OpCode::PROTO);
+  push<uint8_t>(2);
+
+  // All attributes get pushed into a list and their indices saved in the
+  // module def
+  push<OpCode>(OpCode::EMPTY_LIST);
+  push<OpCode>(OpCode::MARK);
+}
+
+void Pickler::finish() {
+  push<OpCode>(OpCode::APPENDS);
+  push<OpCode>(OpCode::STOP);
+}
+
+void Pickler::addIValue(const IValue& ivalue) {
+  // Check if reference ivalue has been saved before
+  const void* ivalue_ptr = getPointer(ivalue);
+  if (ivalue_ptr) {
+    auto memo_entry = memo_.find(ivalue_ptr);
+    if (memo_entry != memo_.end()) {
+      // This value has already been pushed, just do a BINGET
+      pushBinGet(memo_entry->second);
+      return;
+    }
+  }
+
+  if (ivalue.isTensor()) {
+    pushTensor(ivalue);
+  } else if (ivalue.isTuple()) {
+    pushTuple(ivalue);
+  } else if (ivalue.isDouble()) {
+    pushDouble(ivalue);
+  } else if (ivalue.isInt()) {
+    // TODO: use BININT1/BININT2/LONG if possible/necessary
+    AT_ASSERT(ivalue.toInt() <= std::numeric_limits<int32_t>::max());
+    push<OpCode>(OpCode::BININT);
+    push<int32_t>(ivalue.toInt());
+  } else if (ivalue.isBool()) {
+    if (ivalue.toBool()) {
+      push<OpCode>(OpCode::NEWTRUE);
+    } else {
+      push<OpCode>(OpCode::NEWFALSE);
+    }
+  } else if (ivalue.isString()) {
+    pushString(ivalue);
+  } else if (ivalue.isGenericList()) {
+    pushList(ivalue);
+  } else if (ivalue.isGenericDict()) {
+    pushDict(ivalue);
+  } else if (ivalue.isNone()) {
+    push<OpCode>(OpCode::NONE);
+  } else {
+    AT_ERROR("Unknown IValue type for pickling");
+  }
+}
+
+const void* Pickler::getPointer(const IValue& ivalue) {
+  if (ivalue.isGenericDict()) {
+    return &(ivalue.toGenericDictRef());
+  } else if (ivalue.isGenericList()) {
+    return &(ivalue.toGenericListRef());
+  } else if (ivalue.isTuple()) {
+    return &(ivalue.toTuple()->elements());
+  } else if (ivalue.isString()) {
+    return &(ivalue.toStringRef());
+  }
+
+  return nullptr;
+}
+
+void Pickler::pushBinGet(uint32_t memo_id) {
+  if (memo_id <= std::numeric_limits<uint8_t>::max()) {
+    push<OpCode>(OpCode::BINGET);
+    push<uint8_t>(memo_id);
+  } else {
+    // Memoized too many items, issue a LONG_BINGET instead
+    push<OpCode>(OpCode::LONG_BINGET);
+    push<uint32_t>(memo_id);
+  }
```

Comment:
This needs to throw an exception when we exceed 2^32 elements. Otherwise it will silently alias objects!